### PR TITLE
control kwargs

### PR DESCRIPTION
**Pull request by vizhur**
adding utils to handle kwargs to start from

### CODE REVIEW

**Comment by willingc** [comment]
Hi @vizhur. Thanks for working on Papermill. Can you explain why you are making this PR? What is the use case? Why this would be a good approach to address the use case? This will help us evaluate the PR since I don't believe it closes any open issue and would be a new feature request.

**Comment by vizhur** [comment]
@willingc that is regarding https://github.com/nteract/papermill/issues/354
related to the recent changes https://github.com/nteract/papermill/pull/351

File: papermill/utils.py [hunk 1]
```diff
@@ -1,11 +1,31 @@
...
+import warnings
 
 from contextlib import contextmanager
 from functools import wraps
 
 logger = logging.getLogger('papermill.utils')
 
+WARN_ON_ARGS_OVERWRITE = True
+
+def safe_kwargs(caller_args, **callee_args):
```
[the remark this answers is not in the document]
Comment (MSeal) [reply papermill/utils.py]: Probably needs a docstring to explain how it is meant to be used for future developers.

File: papermill/utils.py [hunk 2]
```diff
@@ -1,11 +1,31 @@
...
+        conflicts = set(caller_args) & set(callee_args)
+        if conflicts:
+            args = format('; '.join(['{}={}'.format(key, value)
+                                     for key, value in callee_args.items()]))
+            msg = 'Callee will overwrite caller''s argument(s): {}'.format(args)
+            warnings.warn(msg)
+    return dict(caller_args, **callee_args)
+
+
+def remove_args(args=None, kwargs=None):
```
Comment (MSeal) [inline papermill/utils.py]: `def remove_resvered_args(reserved_args=None, user_kwargs=None):`

File: papermill/utils.py [hunk 3]
```diff
@@ -1,11 +1,31 @@
...
+
+def safe_kwargs(caller_args, **callee_args):
+    if WARN_ON_ARGS_OVERWRITE:
+        conflicts = set(caller_args) & set(callee_args)
+        if conflicts:
+            args = format('; '.join(['{}={}'.format(key, value)
+                                     for key, value in callee_args.items()]))
+            msg = 'Callee will overwrite caller''s argument(s): {}'.format(args)
+            warnings.warn(msg)
+    return dict(caller_args, **callee_args)
```
Comment (MSeal) [inline papermill/utils.py]: Is the intention to overwrite application set argument with user arguments or the other way around?

File: papermill/utils.py (hunk 2 above)
Comment (MSeal) [reply papermill/utils.py]: Is this the inverse of `safe_kwargs`?

File: papermill/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vizhur) [reply papermill/utils.py]: @MSeal if it happens down the stream that can be not user args anymore... It's generic a bit, but it's a util function that doesn't 'suppose' to be called by a user. Anyway, once we agreed on the solution I would be happy to add docstrings

File: papermill/utils.py [hunk 4]
```diff
@@ -1,11 +1,31 @@
...
 
 from contextlib import contextmanager
 from functools import wraps
 
 logger = logging.getLogger('papermill.utils')
 
+WARN_ON_ARGS_OVERWRITE = True
+
+def safe_kwargs(caller_args, **callee_args):
+    if WARN_ON_ARGS_OVERWRITE:
```
[the remark this answers is not in the document]
Comment (vizhur) [reply papermill/utils.py]: there were concerns about warnings.. I'd like to be able to turn off them for various reasons

**Comment by willingc** [comment]
@MSeal In [Python docs](https://docs.python.org/3/reference/compound_stmts.html#function-definitions):

> If the form “**identifier” is present, it is initialized to a new ordered mapping receiving any excess keyword arguments, defaulting to a new empty mapping of the same type. 

I'm still not sure what case we are trying to prevent?? I don't understand why we can't just pass the kwargs.

File: papermill/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MSeal) [reply papermill/utils.py]: Fair. Names were just confusing enough I couldn't reason about what actor is taking precedence.

File: papermill/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MSeal) [reply papermill/utils.py]: Users can catch (or use contexts) to suppress warnings. Python actually has a good built-in control plane for them without us.

If you do want a control switch beyond that it shouldn't be a const in the file as editing that from outside the package would be a bit of an anti-pattern.

**Comment by MSeal** [comment]
@willingc in the simplest form this happens now:

```
def top_level_func(foo, bar=None, **kwargs):
  downstream_func(bar=bar, timeout=50, **kwargs) # TypeError: downstream_func() got multiple values for keyword argument 'timeout'
```

We pass kwargs to multiple downstream callers which get augmented with named defaults which can conflict. I think @betatim 's main complaint was that it's making implicit contracts about reserved keywords in the kwargs. @vizhur was making an attempt to generate a utility to warn when these keywords conflict and continue execution rather than raising.

The fact it's about how we should handle implicit internal contracts when exposing `kwargs` and that we hadn't made a decision about how we'd handle these conflicts meant I pushed for #351 to merge as is and to discuss outside if we wanted to be more explicit in our handling of the problem. One option is to continue with passing **kwargs and let it raise when there's a conflict and not merge anything. But the issue had opinions that we should be a little more aware of potential kwargs conflicts in our plugin system as version bumps occur.

Does that makes the PR context a little clearer?

**Comment by willingc** [comment]
Thank @MSeal for the clarification. I need to do a bit of thinking about this and also PEP 570. I'm trying to balance whether we are overly constraining which over time could be a maintenance headache against being more explicit about behavior (avoiding @betatim's implicit contracts about reserved keywords).

**Comment by MSeal** [comment]
@vizhur Ok got enough of a direction to move forward. Outside of the single-line comments, let's do this:

- Let's change `safe_kwargs` to `merge_kwargs` to make it clearer what the function goal is.
- Then let's add it to the few places where we pass kwargs down and combine with other arguments for the next function call.
- Name kwargs to be the clearer about where these args are being passed (e.g. `engine_kwargs`)
- Add some tests for emitting warnings on merging overriding kwargs (I can help here if setting up the mocks / warnings capture is difficult) 
- Finally we should add to the DEVELOPMENT_GUIDE.md about minimizing passing extra kwargs downstream to functions.

Once the PR has these we can do normal PR review and get a merge + release.

File: papermill/utils.py (hunk 3 above)
Comment (vizhur) [reply papermill/utils.py]: function can overwrite arguments it passes down

File: papermill/utils.py (hunk 2 above)
Comment (vizhur) [reply papermill/utils.py]: kinda, need to remove args from kwargs intended for funtion but not listed in kwargs or named differently, like timeout/execution_timeout

**Comment by vizhur** [comment]
@MSeal @willingc please review the changes

File: papermill/tests/test_engines.py [hunk 5]
```diff
@@ -1,520 +1,520 @@
-import abc
```
Comment (MSeal) [inline papermill/tests/test_engines.py]: Do you know why this file got entirely touched? Was there a newline swap?

**Comment by MSeal** [comment]
Outside of the one file being completely touched this LGTM. Given it's compressed to a few easily read lines of code I'm not too worried about maintenance / complexity burden here. Unless Carol objects or wants any changes I'll plan to merge and release after the test_engines.py comment.

**Comment by vizhur** [comment]
@MSeal I reverted test_engines mess

**Comment by willingc** [comment]
@vizhur @MSeal I appreciate the work that you are doing on this PR and the prior #351. I'm going to approve the PR but wish to open an issue to look at the design around engines and execution. Down the road, we may wish to refactor, but for now this seems to meet Microsoft's needs so let's go ahead and merge.

As next steps, I plan to document in more detail what is happening under the hood #368. After that, we should do a design review re: engines and execution to see what could be improved / clarified.

Thanks and :shipit:

**Comment by willingc** [comment]
Thanks @vizhur :cake: Always great to have new contributors. Welcome!

**Pull request merged**